tests/python/pants_test/build_graph/test_target.py

```diff
@@ -110,6 +112,30 @@ def test_unknown_kwargs(self):
     target = self.make_target('foo:bar', Target, foobar='barfoo')
     self.assertFalse(hasattr(target, 'foobar'))
 
+  def test_tags_applied_from_configured_json(self):
+    with temporary_file(binary_mode=False) as fp:
+      json.dump({
+        'tag_targets_mappings': {
+          'special_tag': ['foo:bar', 'path/to/target:foo', 'path/to/target'],
+          'special_tag2': ['path/to/target:target', '//base:foo']
+        }
+      }, fp)
+      fp.flush()
+
+      options = {Target.TagAssignments.options_scope: {'tag_targets_file': fp.name}}
+      init_subsystem(Target.TagAssignments, options)
+      target1 = self.make_target('foo:bar', Target, tags=['tag1', 'tag2'])
+      target2 = self.make_target('path/to/target:foo', Target, tags=['tag1'])
+      target3 = self.make_target('path/to/target', Target, tags=['tag2'])
+      target4 = self.make_target('//base:foo', Target, tags=['tag3'])
+      target5 = self.make_target('baz:qux', Target, tags=['tag3'])
+
+      self.assertEqual({'tag1', 'tag2', 'special_tag'}, target1.tags)
+      self.assertEqual({'tag1', 'special_tag'}, target2.tags)
+      self.assertEqual({'tag2', 'special_tag', 'special_tag2'}, target3.tags)
+      self.assertEqual({'tag3', 'special_tag2'}, target4.tags)
+      self.assertEqual({'tag3'}, target5.tags)
+
```

Comment:
Things that might be nice to test
- what happens when a target has the special tag already (probably nothing)
- what happens when the tag target mapping points to a non-existent target (probably ignore it--differentiating non-existent vs not in the current closure is probably something the subsystem shouldn't worry about)
- if an invalidly formatted spec is in the source file, what does the error look like (It'd be nice if the error pointed to the JSON file)

Reply:
>what happens when a target has the special tag already (probably nothing)

Nothing, I included `target5` to show nothing changes if things don't match, but can add more cases if it's helpful

>what happens when the tag target mapping points to a non-existent target (probably ignore it--differentiating non-existent vs not in the current closure is probably something the subsystem shouldn't worry about)

It's ignored, added a case for that in the test - though that illustrates that it's possible to define things in the tag mappings that don't do anything, not sure if it'd be preferable to maybe raise an error if something is defined that doesn't match

>if an invalidly formatted spec is in the source file, what does the error look like (It'd be nice if the error pointed to the JSON file)

Just switched it to use a `dict` and `fromfile` instead of JSON, tried things out with a malformed JSON file and a `ParseError` was raised